markov-analysis.py

```diff
@@ -0,0 +1,220 @@
+    mapping of n-tuple (of n prefixes) to some collection that has suffixes with 
+    their respective frequencies
+
+    e.g. 
+
+    (p_1, p_2, ..., p_n) -> ((s_1, f_1), (s_2, f_2), ..., (s_n, f_n))
+    where the order of p matters
+    ideally the f_i's would be normalized but iirc doing that to randomly pick a 
+    float x between 0 and 1 might result in something not random, so i think i 
+    will go with the cumulative sum solution from the preview problem
+    another solution might be to just construct a list of each occurrence 
+
+    simple idea:
+    don't use word frequencies, just a set of suffixes
+    so the dict is a mapping P -> S 
+    (p_1, p_2, ..., p_n) -> (s_1, s_2, ..., s_n)
+
+    after finishing:
+    not sure why i ever thought of using a set of suffixes instead of a list,
+    since that would even preserve relative frequencies!
+'''
+
+def mashup(prefix_length, text_length, *titles):
+    '''
+    Generates random text using two or more books as source material.
+
+    TODO: process_file does not properly skip the Project Gutenberg header
+          (see skip_gutenberg_header for more information)
+
+    prefix_length: int, the number of prefix words to use
+    text_length: int, the length of the random text
+    titles: strings, titles of the source books
+    '''
+    word_list = []
+    for title in titles:
+        # WARNING: skip_header only works for emma.txt, Project Gutenberg books
+        #          do NOT have a standard header format
+        word_list = word_list + process_file(title, skip_header=False)
+    prefixes = prefix_map(word_list, prefix_length)
+    return list_to_string(random_text(prefixes, prefix_length, text_length))
+
+
+def generate_text(prefix_length, text_length, title='emma.txt'):
+    '''
+    Generates random text using the specified number of prefixes.
+
+    prefix_length: int, the number of prefix words to use
+    text_length: int, the length of the random text
+    title: string, title of the source book
+
+    int, int -> str
+    '''
+    word_list = process_file(title, skip_header=True)
+    prefixes = prefix_map(word_list, prefix_length)
+    return list_to_string(random_text(prefixes, prefix_length, text_length))
+
+def list_to_string(word_list):
+    '''
+    Converts a list of words to a string, with spaces between each word and a 
+    period at the end.
+
+    word_list: list
+
+    list -> str
+    '''
+    s = ''
+    for word in word_list:
+        s = s + ('%s ' % word)
+    return (s.strip() + '.')
+
+def random_text(prefix_map, prefix_length, text_length):
+    '''
+    Returns a list of words randomly selected by prefix using the given 
+    prefix map.
+
+    # TODO: This looks messy, is there a better way than changing so many types?
+
+    prefix_map: dict of tuples to lists
+    prefix_length: int, the length (in words) of the sequence of prefixes to use
+    m: int, the length (in words) of the list of random words
+
+    dict, int, int -> list
```

Comment:
why would you want to return a list of strings here, rather than a string?

Reply:
i don't remember my motivation for doing this. i think, since we don't need all the words as a single `string` until the end, i may have wanted to do one of two things:

- delay converting to `string`, in case we ended up processing the text more later

- break up how much any individual function did

but in terms of refactoring this code, returning a `string` would be clearer and simpler.

Reply:
done (now returns `str`)